### PR TITLE
define the interface for the "pyramid"

### PR DESCRIPTION
**Pull request by inabao**
#166

### CODE REVIEW

File: examples/cpp/simple_pyramid.cpp [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <vsag/vsag.h>
+
+#include <iostream>
```
Comment (jiaweizone) [inline examples/cpp/simple_pyramid.cpp:18]: remove unused header file

File: src/dataset_impl.h [hunk 2]
```diff
@@ -30,7 +30,8 @@
 namespace vsag {
 
 class DatasetImpl : public Dataset {
-    using var = std::variant<int64_t, const float*, const int8_t*, const int64_t*>;
+    using var =
+        std::variant<int64_t, const float*, const int8_t*, const int64_t*, const std::string*>;
```
Comment (jiaweizone) [inline src/dataset_impl.h:34]: Could you please add the field comments?

File: examples/cpp/simple_pyramid.cpp (hunk 1 above)
Comment (inabao) [reply examples/cpp/simple_pyramid.cpp:18]: the exmaple need to output the information

File: src/dataset_impl.h (hunk 2 above)
Comment (inabao) [reply src/dataset_impl.h:34]: The explanation of these fields can be found at the interface.

File: src/dataset_impl.h [hunk 3]
```diff
@@ -162,6 +163,20 @@ class DatasetImpl : public Dataset {
...
     }
 
+    DatasetPtr
+    Paths(const std::string* paths) override {
+        this->data_[DATASET_PATHS] = paths;
+        return shared_from_this();
+    }
+
+    const std::string*
+    GetPaths() const override {
```
Comment (jiaweizone) [inline src/dataset_impl.h:175]: Add memory free logic for new `path_` object

Comment (inabao) [reply src/dataset_impl.h:175]: done

**Review by jiaweizone — APPROVED** [review]
LGTM

File: examples/cpp/simple_pyramid.cpp [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+    const std::vector<std::string> level2 = {"d", "e"};
+    const std::vector<std::string> level3 = {"f", "g", "h"};
+
+    std::random_device rd;
+    std::mt19937 mt(rd());
+
+    std::vector<std::string> selected_levels;
+
+    if (is_full) {
+        selected_levels.push_back(level1[rand() % level1.size()]);
```
Comment (LHT129) [inline examples/cpp/simple_pyramid.cpp]: emplace_back

File: src/index/pyramid.h [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+#include "vsag/vsag.h"
+
+namespace vsag {
+
+class Pyramid : public Index {
+public:
+    Pyramid() {
+    }
+
+    ~Pyramid() = default;
```
Comment (LHT129) [inline src/index/pyramid.h]: override

File: src/index/pyramid.h [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+    Deserialize(const ReaderSet& reader_set) override;
+
+    int64_t
+    GetNumElements() const override;
+
+    int64_t
+    GetMemoryUsage() const override;
+
+private:
+};
```
Comment (LHT129) [inline src/index/pyramid.h:78]: nothing?

File: src/index/pyramid.h [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include "vsag/vsag.h"
+
```
Comment (LHT129) [inline src/index/pyramid.h:19]: suggest include the headers only used

File: examples/cpp/simple_pyramid.cpp [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+    int64_t dim = 128;
+
+    // prepare ids and vectors
+    auto ids = new int64_t[num_vectors];
+    auto vectors = new float[dim * num_vectors];
+    auto paths = new std::string[num_vectors];
+
+    std::mt19937 rng;
+    rng.seed(47);
+    std::uniform_real_distribution<> distrib_real;
```
Comment (LHT129) [inline examples/cpp/simple_pyramid.cpp]: add template float

**Review by LHT129 — CHANGES_REQUESTED** [review]

File: examples/cpp/simple_pyramid.cpp (hunk 4 above)
Comment (inabao) [reply examples/cpp/simple_pyramid.cpp]: done

File: src/index/pyramid.h (hunk 5 above)
Comment (inabao) [reply src/index/pyramid.h]: done

File: src/index/pyramid.h (hunk 6 above)
Comment (inabao) [reply src/index/pyramid.h:78]: done

File: src/index/pyramid.h (hunk 7 above)
Comment (inabao) [reply src/index/pyramid.h:19]: done

File: examples/cpp/simple_pyramid.cpp (hunk 8 above)
Comment (inabao) [reply examples/cpp/simple_pyramid.cpp]: done

File: examples/cpp/simple_pyramid.cpp (hunk 4 above)
Comment (LHT129) [reply examples/cpp/simple_pyramid.cpp]: not fix all

Comment (inabao) [reply examples/cpp/simple_pyramid.cpp]: done again

**Review by LHT129 — APPROVED** [review]
LGTM

**Pull request merged**